cmd/load-generator/boulder-calls.go

```diff
@@ -0,0 +1,482 @@
+	if err != nil {
+		return err
+	}
+	ctx.ns = &nonceSource{s: s}
+	signer.SetNonceSource(ctx.ns)
+
+	// create the registration object
+	var regStr []byte
+	if s.email != "" {
+		regStr = []byte(fmt.Sprintf(`{"resource":"new-reg","contact":["mailto:%s"]}`, s.email))
+	} else {
+		regStr = plainReg
+	}
+	// build the JWS object
+	requestPayload, err := s.signWithNonce("/acme/new-reg", true, regStr, signer)
+	if err != nil {
+		return fmt.Errorf("/acme/new-reg, sign failed: %s", err)
+	}
+
+	nStarted := time.Now()
+	resp, err := s.post(fmt.Sprintf("%s/acme/new-reg", s.apiBase), requestPayload, ctx.ns)
+	nFinished := time.Now()
+	nState := "good"
+	defer func() { s.callLatency.Add("POST /acme/new-reg", nStarted, nFinished, nState) }()
+	if err != nil {
+		nState = "error"
+		return fmt.Errorf("/acme/new-reg, post failed: %s", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 201 {
+		body, err := ioutil.ReadAll(resp.Body)
+		nState = "error"
+		if err != nil {
+			// just fail
+			return fmt.Errorf("/acme/new-reg, bad response: %s", body)
+		}
+		return fmt.Errorf("/acme/new-reg, bad response status %d: %s", resp.StatusCode, body)
+	}
+
+	// get terms
+	links := resp.Header[http.CanonicalHeaderKey("link")]
+	terms := ""
+	for _, l := range links {
+		if strings.HasSuffix(l, ">;rel=\"terms-of-service\"") {
+			terms = l[1 : len(l)-len(">;rel=\"terms-of-service\"")]
+			break
+		}
+	}
+
+	// agree to terms
+	regStr = []byte(fmt.Sprintf(`{"resource":"reg","agreement":"%s"}`, terms))
+
+	// build the JWS object
+	requestPayload, err = s.signWithNonce("/acme/reg", false, regStr, signer)
+	if err != nil {
+		return fmt.Errorf("/acme/reg, sign failed: %s", err)
+	}
+
+	tStarted := time.Now()
+	resp, err = s.post(resp.Header.Get("Location"), requestPayload, ctx.ns)
+	tFinished := time.Now()
+	tState := "good"
+	defer func() { s.callLatency.Add("POST /acme/reg/{ID}", tStarted, tFinished, tState) }()
+	if err != nil {
+		tState = "error"
+		return fmt.Errorf("/acme/reg, post failed: %s", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 202 {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			// just fail
+			tState = "error"
+			return err
+		}
+		tState = "error"
+		return fmt.Errorf("/acme/reg, bad response status %d: %s", resp.StatusCode, body)
+	}
+
+	ctx.reg = &registration{key: signKey, signer: signer}
+	s.addRegistration(ctx.reg)
+	return nil
+}
+
+func newAuthorization(s *State, ctx *context) error {
+	// generate a random(-ish) domain name,  will cause some multiples but not enough to make rate limits annoying!
+	n := time.Now().UnixNano()
+	b := new(bytes.Buffer)
+	binary.Write(b, binary.LittleEndian, n)
+	randomDomain := fmt.Sprintf("%x.%s", sha1.Sum(b.Bytes()), s.domainBase)
+
+	// create the registration object
+	initAuth := fmt.Sprintf(`{"resource":"new-authz","identifier":{"type":"dns","value":"%s"}}`, randomDomain)
+
+	// build the JWS object
+	getNew := false
+	if mrand.Intn(1) == 0 {
```

Comment:
It might be worth a comment about the purpose of this line. I don't use `math.rand` much and had to look up what `Intn(1)` was doing to understand this was a coin flip to decide whether to set `getNew`.